### ISSUE TITLE
Deep links are broken on iOS

### ISSUE THREAD

**Issue by marcaaron** [Expensify/App#15158]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Install the New Expensify app
2. Sign in and close app
3. Follow a deep link like https://new.expensify.com/settings

## Expected Result:

App should open to the settings page

## Actual Result:

Safari or other default web browser opens

## Workaround:

Open the app directly

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [x] iOS / native
- [ ]  iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** v1.2.71-1
**iOS Version:** 16.3.1 
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
**Expensify/Expensify Issue URL:**
**Issue reported by:** @shawnborton
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1676413158858859

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @arielgreen (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by mvtglobally** [comment]
@marcaaron we had an old Issue on this which was closed. There was a bigger discussion. I'll try to link it in a bit

**Comment by mvtglobally** [comment]
@marcaaron @arielgreen 

https://github.com/Expensify/App/issues/13882#issuecomment-1370603008

**Comment by marcaaron** [comment]
Thanks @mvtglobally (great memory)! 

I'll leave a comment over there. They seem like the same issue so if anything we will keep this one open and leave that one closed.

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01d64bfcd13e8d5909

**Comment by MelvinBot** [comment]
Current assignee @arielgreen is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @parasharrajat (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @NikkiWines (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by bjin9** [comment]
Is it still open for new ones?

**Comment by allroundexperts** [comment]
@NikkiWines @parasharrajat This seems to be a feature request. Did it ever worked on iOS? As far as I can see, `DeepLinkWrapper` was never implemented for native platforms. It is just done for Desktop.

**Comment by allroundexperts** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Deep links do not work on iOS.

### What is the root cause of that problem?
The root cause of this issue is the feature not implemented for iOS and Android as seen in this file `src/components/DeeplinkWrapper/index.website.js`. The links themselves `new-expensify://route` work fine on apps when opened in browsers. However, we're restricting the app to open the links on mobile here `src/components/DeeplinkWrapper/index.website.js`.

### What changes do you think we should make in order to solve the problem?

We can simply remove the conditions that are restricting the code to run on mobile browsers in `src/components/DeeplinkWrapper/index.website.js` file. I'm working on the code changes and will post it here soon.

### What alternative solutions did you explore? (Optional)
None

**Comment by cubuspl42** [comment]
Proposal:
* On the web application side:
  * Add static `apple-app-site-association` manifest file that reflects [deeplinkRoutes.js](https://github.com/Expensify/App/blob/137b9477843789408969c71c41a4cf87e45c9a4c/src/components/DeeplinkWrapper/deeplinkRoutes.js) content
* On the iOS app project side:
  * Declare necessary entitlements
* On the React Native (JavaScript) side:
  * Generalize the logic currently contained in [ReportUtils.openReportFromDeepLink](https://github.com/Expensify/App/blob/a5b9c9ba88fa0634a963bd248152b79e3006bd6d/src/libs/ReportUtils.js#L1493) adding support for other types of deep links. Let's say, creating a new helper named `LinkingUtils.openFromDeepLink`.
  
Also, some minor cofiguration will need to be done on the App Store Connect side.

**Comment by s77rt** [comment]
This may be due to https://new.expensify.com/.well-known/apple-app-site-association not setting the correct `content-type` header value, it should be `application/json`.

**Comment by s77rt** [comment]
@parasharrajat The `apple-app-site-association` file is a JSON file and I think iOS expects the server to return the correct content type otherwise it will fail to consider checking for deep links (universal links). Not much of a proposal as I didn't go deep to understand the root cause since the incorrect content-type may be a blocker, so it's worth the fix anyway before proceeding with frontend proposals (if the issue presists).

**Comment by marcochavezf** [comment]
IIRC this worked fine for iOS previously, @mvtglobally @kbecciv do you know if this is a regression?

**Comment by marcochavezf** [comment]
Moving the convo to Slack [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1677180758585259) for broader visibility

**Comment by NikkiWines** [comment]
Not overdue, in discussion

**Comment by mvtglobally** [comment]
@marcochavezf rhere was an older issue about it. Doesn't seem like a new issue https://github.com/Expensify/App/issues/13882
But not sure if it's a regression due to another Pr

**Comment by MelvinBot** [comment]
@NikkiWines, @parasharrajat, @arielgreen Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by marcochavezf** [comment]
Commented in the thread

**Comment by marcochavezf** [comment]
Chatted with @NikkiWines, I'm going to take this issue

**Comment by MelvinBot** [comment]
@marcochavezf @parasharrajat @arielgreen this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by MelvinBot** [comment]
Current assignee @parasharrajat is eligible for the Internal assigner, not assigning anyone new.

**Comment by arielgreen** [comment]
Moved internal, deleting Upwork job

**Comment by marcochavezf** [comment]
Looked for help in [infra](https://expensify.slack.com/archives/C094TGUTZ/p1677891723144469) to set the `content-type` header.

**Comment by marcochavezf** [comment]
Created an [issue](https://github.com/Expensify/Expensify/issues/266451) (for infra) to set the correct MIME type

**Comment by arielgreen** [comment]
Related PR https://github.com/Expensify/App/pull/15708

**Comment by marcochavezf** [comment]
It was [reverted](https://github.com/Expensify/App/pull/15727) because it was breaking the file on iOS. More details [here](https://expensify.slack.com/archives/C04QEB4MJEQ/p1678228607435939).

**Comment by marcochavezf** [comment]
Yeah, this is an issue with the deployment process for web. A [fix](https://github.com/Expensify/App/pull/15743) for this was merged a few mins ago.

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.81-1 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/15743


If no regressions arise, payment will be issued on 2023-03-17. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@parasharrajat / @marcochavezf] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@parasharrajat / @marcochavezf] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@parasharrajat / @marcochavezf] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@arielgreen] Determine if we should create a regression test for the bug.
- [ ] [@parasharrajat] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to the [appropriate location](https://github.com/Expensify/App/blob/main/contributingGuides/TestRail_Navigation_C%2B.md) to ensure the same bug will not reach production again.
- [ ] [@arielgreen] Review the proposed regression test steps and location.
- [ ] [@arielgreen] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by MelvinBot** [comment]
Looks like something related to `react-navigation` may have been mentioned in this issue discussion.

As a reminder, please make sure that all proposals are not workarounds and that any and all attempt to fix the issue holistically have been made before proceeding with a solution. Proposals to change our `DeprecatedCustomActions.js` files should not be accepted.

Feel free to drop a note in [#expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q) with any questions.

**Comment by marcochavezf** [comment]
Oops, the [PR](https://github.com/Expensify/App/pull/15743) to change the content-type for the AASA file was merged, but the deep links are not opening the installed iOS app instead of the browser.

**Comment by marcochavezf** [comment]
The universal link diagnostics and the AASA validator indicate that everything is fine:

https://user-images.githubusercontent.com/6829422/224510009-236894b0-88eb-46ec-9d02-2d0b891f1d57.MOV

But links to production and staging are still opening the broswer 

e.g.

https://new.expensify.com/r/3504895439653267

https://staging.new.expensify.com/r/3504895439653267

**Comment by s77rt** [comment]
It may be worth to check if this is related to cache, I think apple check those urls once a week but not really sure

**Comment by s77rt** [comment]
@marcochavezf Can you please remove the "HOLD" from the title, this is not fixed yet.

**Comment by marcochavezf** [comment]
> It may be worth to check if this is related to cache, I think apple check those urls once a week but not really sure

Yeah, that was my first thought, but it has already passed a week since the apple-app-site-association file was updated. Probably we'd need to wait for a few days more.

**Comment by MelvinBot** [comment]
@marcochavezf, @parasharrajat, @arielgreen Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by marcochavezf** [comment]
Update [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1679412576275409?thread_ts=1677180758.585259&cid=C01GTK53T8Q). It seems that we'd need to wait until the other change is deployed to production and test again.

**Comment by arielgreen** [comment]
@marcochavezf is this still held?

**Comment by marcochavezf** [comment]
The other files were already updated on production and staging, but the deep links are not still working.

<img width="973" alt="Screenshot 2023-03-27 at 10 26 51" src="https://user-images.githubusercontent.com/6829422/227915953-0f9ef2cc-df01-45ed-96e8-e87e17a4f000.png">

<img width="1091" alt="Screenshot 2023-03-27 at 10 26 28" src="https://user-images.githubusercontent.com/6829422/227916027-137b0562-f11f-4bc4-88b0-eaa67e89f27f.png">

**Comment by marcochavezf** [comment]
Looking for more ideas of what could be the issue [here](https://expensify.slack.com/archives/C03UK30EA1Z/p1679916728333649?thread_ts=1678567590.302789&cid=C03UK30EA1Z).

**Comment by mczernek** [comment]
I did some research on that and found some interesting things.

It was extremely difficult for me to get any one app to handle universal links. However I discovered that most domains uses different formats of AASA file than we do. Great example (and working one!) is YouTube:

```
{
    "applinks": {
        "apps": [],
        "details": [
            {
                "appID": "EQHXZ8M8AV.com.google.ios.youtube.dev",
                "paths": [
                    "NOT /yt/*",
						// Whole bunch of NOTs here,
                    "*"
                ]
            },
            // tonns of other appIDs
		]
	}
}
```

The differences to our file are:
* instead of `components` they have `paths` which is array of strings
* they don't have `webcredentials` key within `applinks`

There are other domains that follow similar format (pretty much every one I checked) but with `webcredentials` in it.

Interestingly youtube.com is the only one that Diagnostics in Settings finds correct (apart from us, but unlike us YouTube actually works).

**Comment by mczernek** [comment]
After some more investigation I have one more thought to be tested.

In fact, iOS does not query our associated file directly. Instead, it uses apple-cdn by getting from url such like `https://app-site-association.cdn-apple.com/a/v1/new.expensify.com`

For some reason it does return 404 for our domain, while it works for pretty much every other I tested. My guess would be that there is something wrong with our response when serving associated-domains file. In fact, there are quite some headers and two cookies being returned. Can we try simplify that as much as possible and leave only relevant headers and no cookies. For reference, this is set of headers returned by branch for one of apps:

```
content-type: application/json; charset=utf-8
content-length: 1362
vary: Accept-Encoding
server: openresty
date: Tue, 28 Mar 2023 14:34:49 GMT
strict-transport-security: max-age=31536000; includeSubDomains
x-cache: Miss from cloudfront
via: 1.1 2e4b77c76f89825e36f12179cf1b33ea.cloudfront.net (CloudFront)
x-amz-cf-pop: WAW51-P1
x-amz-cf-id: 8lgANq-rSQNzbOs4GuRIniAxHrXVjNrQCw9dVy48ZLPKPkdzc6OM-g==
```

**Comment by marcochavezf** [comment]
Thanks for the insights @mczernek! We're discussing with the infra team the headers issue.

**Comment by marcochavezf** [comment]
After more investigation seems that our firewall is blocking Apple's crawler from pulling the AASA file from our servers. [Infra](https://expensify.slack.com/archives/C094TGUTZ/p1680168395398899) already updated the rule. So I think we should only need to wait a few days to corroborate if the apple-cdn is updated.

**Comment by marcochavezf** [comment]
It's working now! 

https://user-images.githubusercontent.com/6829422/228837759-78cf462a-a40b-4d6b-90f4-05a4f8bee6c1.MP4

**Comment by mczernek** [comment]
I reinstalled the app, restated the phone and it worked 🎉 
Good job!

**Comment by marcochavezf** [comment]
For future reference, if someone has a similar problem in the future, the problem was the firewall.

We were getting this error from the headers of the apple cdn `Apple-Failure-Reason: SWCERR00101 Bad HTTP Response: 403 Forbidden`:

<img width="1232" alt="Screenshot 2023-03-29 at 15 04 05" src="https://user-images.githubusercontent.com/6829422/228839714-c04b914f-e885-4ea8-b5c7-b76dd19485ab.png">

And someone else that had the same error, posted in the apple developer forum that the problem was their firewall configuration: https://developer.apple.com/forums/thread/675324?answerId=668081022#668081022

**Comment by marcochavezf** [comment]
I will close this one since it's working on staging (app from TestFlight) and production. Thank you all so much for your help!
